src/vm_workload.py

```diff
@@ -5,131 +5,207 @@
 """KafkaSnap class and methods."""
 
 import logging
+import os
+import secrets
+import shutil
+import string
 import subprocess
-from typing import List
 
 from charms.operator_libs_linux.v0 import apt
 from charms.operator_libs_linux.v1 import snap
 from tenacity import retry
 from tenacity.retry import retry_if_not_result
 from tenacity.stop import stop_after_attempt
 from tenacity.wait import wait_fixed
+from typing_extensions import override
 
-from literals import CHARMED_KAFKA_SNAP_REVISION, SNAP_NAME
+from core.workload import PathsBase, WorkloadBase
+from literals import CHARMED_KAFKA_SNAP_REVISION, PATHS, SNAP_NAME
 
 logger = logging.getLogger(__name__)
 
 
-class KafkaSnap:
-    """Wrapper for performing common operations specific to the Kafka Snap."""
+class KafkaPaths(PathsBase):
```

Comment:
**todo** This seems to be a bit of difference with respect to Zookeeper. Also, I'm afraid that the position here creates a bit of a mess with the dependency import ordering. 

In Zookeeper, this definition is in `core/workload.py` which is fine, because these I would believe it applies to Kafka for both VM and K8s. Placing this definition here, also creates a bit of harder dependency tree, since: general workloads uses this class, indeed this class is also used on substrate-independent implementation (e.g. managers) but it is found on "substrate specific" parts. 

In general, we should have some sort of constraints on importing as outlined [here](https://github.com/canonical/zookeeper-operator/pull/103#pullrequestreview-1807866225). This would create an extremely clean and ordered structure, preventing cross-relations between modules.

Reply:
I'm not sure to follow. `KafkaPaths` is exclusively used within `KafkaWorkload`. `core/workload.py` has evolved in my view to work as "common" and it's only used on the substrate specific file (`src/workload.py`). But I believe it looks much more clean to have it like this:
```pyhton
src/core/workload.py
 - PathsBase
 - WorkloadBase

src/workload.py
 - KafkaPaths(PathsBase)
 - KafkaWorkload(WorkloadBase)
```

Note that KafkaPaths() is ever so slightly different between k8s and vm, with k8s needing `java_home`. So the logical way to set it up is to have KafkaPaths next to KafkaWorkload.